source/adaptive_load/adaptive_load_controller_impl.cc

```diff
@@ -0,0 +1,195 @@
+#include "adaptive_load/adaptive_load_controller_impl.h"
+
+#include <chrono>
+
+#include "envoy/common/exception.h"
+#include "envoy/config/core/v3/base.pb.h"
+
+#include "nighthawk/adaptive_load/adaptive_load_controller.h"
+#include "nighthawk/adaptive_load/metrics_plugin.h"
+#include "nighthawk/adaptive_load/scoring_function.h"
+#include "nighthawk/adaptive_load/step_controller.h"
+
+#include "external/envoy/source/common/common/logger.h"
+#include "external/envoy/source/common/common/statusor.h"
+#include "external/envoy/source/common/protobuf/protobuf.h"
+
+#include "api/adaptive_load/adaptive_load.pb.h"
+#include "api/adaptive_load/benchmark_result.pb.h"
+#include "api/adaptive_load/metric_spec.pb.h"
+#include "api/client/options.pb.h"
+#include "api/client/output.pb.h"
+#include "api/client/service.grpc.pb.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/status/status.h"
+#include "absl/strings/str_join.h"
+#include "adaptive_load/metrics_plugin_impl.h"
+#include "adaptive_load/plugin_loader.h"
+
+namespace Nighthawk {
+
+namespace {
+
+using namespace Envoy; // for Envoy::Logger
```

Comment:
Do we still have to do this? I was hoping that with https://github.com/envoyproxy/envoy/pull/12827 and https://github.com/envoyproxy/envoy/pull/12706 the need for this is gone.

Reply:
Ahh I had not been following that. Removed `using namespace Envoy;` and it still works.

> Generally LGTM, it's quite cool to see everything come together. I feel it reads like poetry :-) except for the special way we treat duration, which still feels a tiny bit off; maybe a code level comment can help here. Somehow I'm also wondering if duration isn't something we should allow implementing as a parameter, like qps. E.g. to allow one to research a question like "how does test duration influence convergence speed at a steady rps".
> Left a few small other remarks.

Shall I compare thee to a summer's *INSERT DURATION HERE*? ;-)

Opened #539 to discuss this. I agree duration sticks out like a sore thumb in this API, and it may be possible to generalize it with minimal cost.

Reply:
Haha nice one :-) Thanks for adding the comment & filing that issue. Resolving.